File: gui/super.cpp
#pragma once
#include "super.h"
#include <QMessageBox>


void gui_error_super(const int level, const char* msg)
{
	if (level > 0)
	{
		QMessageBox::critical(nullptr, "ERROR", QString(msg));
		QMainWindow* main_window = static_cast<QMainWindow*>(get_main_window_ptr());
		main_window->close();
		exit(level);
	}
	else
		throw std::runtime_error(msg);
}

void gui_warning_super(const char* msg)
{
	QMessageBox::information(nullptr, "INFORMATION", QString(msg));
}

void gui_msg_super(const char* msg)
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->display_output(QString(msg));
}

int gui_confirm_super(const char* msg)
{
	QMessageBox::StandardButton answer = QMessageBox::question(nullptr, "CONFIRM", QString(msg));
	if (answer == QMessageBox::Ok) return 1;
	else return 0;
}

void gui_panic_super()
{
	QString msg("Attempting to save Data saved in ws.*files\n");
	msg += "To restart iode with the last data, type :\n";
	msg += "    iode -ws";
	QMessageBox::critical(nullptr, "SWAP PANIC", msg);
	K_end_ws(1);
	QMainWindow* main_window = static_cast<QMainWindow*>(get_main_window_ptr());
	main_window->close();
	exit(2);
}

// ========================================================================= //
// See C API -> report/engine/b_rep_super.c and report/engine/b_rep_syntax.c //
// ========================================================================= //

int gui_PrintObjDef_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_print_objects_definitions_dialog();
	return 0;
}

int gui_ViewPrintGr_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_graphs_tables_dialog();
	return 0;
}

int gui_ViewPrintTbl_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_print_tables_dialog();
	return 0;
}

int gui_ViewByTbl_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_edit_tables_variables_dialog();
	return 0;
}

int gui_DataSearch_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_search_text_dialog();
	return 0;
}

int gui_DataDuplicate_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_duplicate_objects_dialog();
	return 0;
}

int gui_DataList_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_file_contents_dialog();
	return 0;
}

int gui_DataCompare_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_file_compare_dialog();
	return 0;
}

int gui_DataCalcLst_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_list_calculus_dialog();
	return 0;
}

int gui_DataListSort_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_sort_and_calculus_dialog();
	return 0;
}

int gui_DataEditGraph_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_graphs_variables_dialog();
	return 0;
}

int gui_DataScan_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_scan_objects_dialog();
	return 0;
}

int gui_StatUnitRoot_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	UnitRootDialog dialog(main_window);
	return dialog.exec();
}

int gui_WsLoad_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_project();
	return 0;
}

int gui_WsSave_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->save_all_tabs();
	return 0;
}

int gui_WsMerge_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_merge_into_workspace_dialog();
	return 0;
}

int gui_WsDescr_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_change_workspace_descriptions_dialog();
	return 0;
}

int gui_WsCopy_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_copy_into_workspace_dialog();
	return 0;
}

int gui_WsClear_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->clear_workspace();
	return 0;
}

int gui_WsSample_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_change_variables_sample_dialog();
	return 0;
}

int gui_WsExtrapolate_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_extrapolate_variables_dialog();
	return 0;
}

int gui_WsHtoL_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_high_to_low_dialog();
	return 0;
}

int gui_WsLtoH_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_low_to_high_dialog();
	return 0;
}

int gui_WsSeasonAdj_super()
{
	QString lec = "";
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	DynAdjustmentDialog dialog(lec, main_window);
	return dialog.exec();
}

int gui_WsTrend_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_trend_correction_dialog();
	return 0;
}

int gui_ModelSimulate_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_compute_simulation_dialog();
	return 0;
}

int gui_ModelCompile_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_compute_model_dialog();
	return 0;
}

int gui_IdtExecute_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_compute_identities_dialog();
	return 0;
}

int gui_EqsEstimate_super()
{
	IodeTabWidget* tab_widget = static_cast<IodeTabWidget*>(get_tabs_widget_ptr());
	tab_widget->openEditDialog(IodeType::EQUATIONS);
	return 0;
}

int gui_XodeRuleImport_super()
{
	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	main_window->open_import_variables_dialog();
	return 0;
}

int gui_view_tbl(TBL* tbl, char* smpl, char* name)
{
	int pos = K_find(KT_WS, const_cast<char*>(name));
	bool table_found = pos >= 0;

	QString tableName = QString(name);
	QString gsample = QString(smpl);
	int nb_decimals = 6;

	if(!table_found)
	{
		// temporary add the passed IODE table to the global database of tables
		pos = K_add(KT_WS, name, tbl);
		if(pos < 0)
		{
			QMessageBox::warning(nullptr, "WARNING", "Could not compute table '" + QString(name) + 
				"' with sample '" + gsample + "'");
			return QDialog::Rejected;
		}
	}

	MainWindow* main_window = static_cast<MainWindow*>(get_main_window_ptr());
	ComputedTableNumericalDialog view(tableName, gsample, nb_decimals, QString(""), main_window);
	int res = view.exec();

	if(!table_found)
	{
		// remove the temporary added IODE table from the global database
		K_del(KT_WS, pos);
	}

	return res;
}

// ============================================================================== //

void gui_assign_super_API()
{
	IODE_assign_super_API();
	kerror_super = gui_error_super;
	kwarning_super = gui_warning_super;
	kmsg_super = gui_msg_super;
	kconfirm_super = gui_confirm_super;
	kpanic_super = gui_panic_super;

	SB_PrintObjDef_super    = gui_PrintObjDef_super;
	SB_ViewPrintGr_super    = gui_ViewPrintGr_super;
	SB_ViewPrintTbl_super   = gui_ViewPrintTbl_super;
	SB_ViewByTbl_super      = gui_ViewByTbl_super;
	SB_DataSearch_super     = gui_DataSearch_super;
	SB_DataDuplicate_super  = gui_DataDuplicate_super;
	SB_DataList_super       = gui_DataList_super;
	SB_DataCompare_super    = gui_DataCompare_super;
	SB_DataCalcLst_super    = gui_DataCalcLst_super;
	SB_DataListSort_super   = gui_DataListSort_super;
	SB_DataEditGraph_super  = gui_DataEditGraph_super;
	SB_DataScan_super       = gui_DataScan_super;
	SB_StatUnitRoot_super   = gui_StatUnitRoot_super;
	SB_WsLoad_super         = gui_WsLoad_super;
	SB_WsSave_super         = gui_WsSave_super;
	SB_WsMerge_super        = gui_WsMerge_super;
	SB_WsDescr_super        = gui_WsDescr_super;
	SB_WsCopy_super         = gui_WsCopy_super;
	SB_WsClear_super        = gui_WsClear_super;
	SB_WsSample_super       = gui_WsSample_super;
	SB_WsExtrapolate_super  = gui_WsExtrapolate_super;
	SB_WsHtoL_super         = gui_WsHtoL_super;
	SB_WsLtoH_super         = gui_WsLtoH_super;
	SB_WsSeasonAdj_super    = gui_WsSeasonAdj_super;
	SB_WsTrend_super        = gui_WsTrend_super;
	SB_ModelSimulate_super  = gui_ModelSimulate_super;
	SB_ModelCompile_super   = gui_ModelCompile_super;
	SB_IdtExecute_super     = gui_IdtExecute_super;
	SB_EqsEstimate_super    = gui_EqsEstimate_super;
	SB_XodeRuleImport_super = gui_XodeRuleImport_super;

	T_view_tbl_super        = gui_view_tbl;
}

File: gui/iode_gui/super.py
import sys
from iode import (IodeType, comments, equations, identities, lists, scalars, 
				  tables, variables)
from iode.util import register_super_function, c_api_error_as_exception
from iode.super import skip_pause, skip_msg_box

from PySide6.QtWidgets import QMainWindow, QMessageBox

from iode_gui.main_window import MainWindow
from iode_gui.main_widgets.tab_widget.tab_widget import IodeTabWidget
from iode_gui.iode_objs.misc.unit_root import UnitRootDialog
from iode_gui.iode_objs.misc.dyn_adjustment import DynAdjustmentDialog
from iode_gui.tabs.iode_objs.tab_computed_table import ComputedTableDialog


__all__ = ['gui_assign_super']

main_window: MainWindow = None

def gui_assign_super(arg_main_window: MainWindow):
	global main_window

	if not isinstance(arg_main_window, MainWindow):
		QMessageBox.critical(None, "ERROR", "The main window is not initialized yet.")
		return
	
	main_window = arg_main_window
	c_api_error_as_exception(True)

@register_super_function('error')
def error_super_GUI(level: int, msg: str) -> int:
	if level > 0:
		QMessageBox.critical(None, "ERROR", "Error: An unexpected error occurred!\n"
							 "Will exit the program now")
		if main_window is not None:
			main_window.close()
		sys.exit(level)
		return -1
	else:
		main_window.display_output(f"ERROR -> {msg}")
		QMessageBox.critical(None, "ERROR", msg)
		return 0
	
@register_super_function('warning')
def warning_super_GUI(msg: str):
	main_window.display_output(f"WARNING -> {msg}")
	QMessageBox.warning(None, "WARNING", msg)

@register_super_function('message')
def msg_super_GUI(msg: str):
	if main_window is not None:
		main_window.display_output(msg)
	else:
		print(msg)

@register_super_function('confirm')
def confirm_super_GUI(msg: str) -> int:
	answer: QMessageBox.StandardButton = QMessageBox.question(None, "CONFIRM", msg)
	return 1 if answer == QMessageBox.StandardButton.Ok else 0

@register_super_function('panic')
def panic_super_GUI():
	msg: str = "Attempting to save Data saved in ws.*files\n"
	comments.save('ws')
	equations.save('ws')
	identities.save('ws')
	lists.save('ws')
	scalars.save('ws')
	tables.save('ws')
	variables.save('ws')

	msg += "To restart iode with the last data, type :\n"
	msg += "    iode -ws"
	QMessageBox.critical(None, "SWAP PANIC", msg)
	main_window.close()
	sys.exit(msg)

@register_super_function('pause')
def pause_super_GUI():
	QMessageBox.information(None, "PAUSE", "Click OK to continue")

@register_super_function('msgbox')
def msgbox_super_GUI(title: str, msg: str) -> int:
	QMessageBox.information(None, title, msg)
	return 1

# ========================================================================= #
# See C API -> report/engine/b_rep_super.c and report/engine/b_rep_syntax.c #
# ========================================================================= #

def check_main_window_exists(func):
	def wrapper(*args, **kwargs):
		if main_window is not None:
			return func(*args, **kwargs)
		else:
			name = func.__name__.replace('_super_GUI', '')
			QMessageBox.critical(None, "ERROR", f"Cannot run '{name}'. The graphical " 
						         "user interface has not been initialized yet.")
	return wrapper

@register_super_function('PrintObjDef')
@check_main_window_exists
def PrintObjDef_super_GUI() -> int:
	main_window.open_print_objects_definitions_dialog()
	return 0

@register_super_function('ViewPrintGr')
@check_main_window_exists
def ViewPrintGr_super_GUI() -> int:
	main_window.open_graphs_tables_dialog()
	return 0

@register_super_function('ViewPrintTbl')
@check_main_window_exists
def ViewPrintTbl_super_GUI() -> int:
	main_window.open_print_tables_dialog()
	return 0

@register_super_function('ViewByTbl')
@check_main_window_exists
def ViewByTbl_super_GUI() -> int:
	main_window.open_edit_tables_variables_dialog()
	return 0

@register_super_function('DataSearch')
@check_main_window_exists
def DataSearch_super_GUI() -> int:
	main_window.open_search_text_dialog()
	return 0

@register_super_function('DataDuplicate')
@check_main_window_exists
def DataDuplicate_super_GUI() -> int:
	main_window.open_duplicate_objects_dialog()
	return 0

@register_super_function('DataList')
@check_main_window_exists
def DataList_super_GUI() -> int:
	main_window.open_file_contents_dialog()
	return 0

@register_super_function('DataCompare')
@check_main_window_exists
def DataCompare_super_GUI() -> int:
	main_window.open_file_compare_dialog()
	return 0

@register_super_function('DataCalcLst')
@check_main_window_exists
def DataCalcLst_super_GUI() -> int:
	main_window.open_list_calculus_dialog()
	return 0

@register_super_function('DataListSort')
@check_main_window_exists
def DataListSort_super_GUI() -> int:
	main_window.open_sort_and_calculus_dialog()
	return 0

@register_super_function('DataEditGraph')
@check_main_window_exists
def DataEditGraph_super_GUI() -> int:
	main_window.open_graphs_variables_dialog()
	return 0

@register_super_function('DataScan')
@check_main_window_exists
def DataScan_super_GUI() -> int:
	main_window.open_scan_objects_dialog()
	return 0

@register_super_function('StatUnitRoot')
@check_main_window_exists
def StatUnitRoot_super_GUI() -> int:
	dialog = UnitRootDialog(main_window)
	dialog.open()
	return 0

@register_super_function('WsLoad')
@check_main_window_exists
def WsLoad_super_GUI() -> int:
	main_window.open_project()
	return 0

@register_super_function('WsSave')
@check_main_window_exists
def WsSave_super_GUI() -> int:
	main_window.save_all_tabs()
	return 0

@register_super_function('WsMerge')
@check_main_window_exists
def WsMerge_super_GUI() -> int:
	main_window.open_merge_into_workspace_dialog()
	return 0

@register_super_function('WsDescr')
@check_main_window_exists
def WsDescr_super_GUI() -> int:
	main_window.open_change_workspace_descriptions_dialog()
	return 0

@register_super_function('WsCopy')
@check_main_window_exists
def WsCopy_super_GUI() -> int:
	main_window.open_copy_into_workspace_dialog()
	return 0

@register_super_function('WsClear')
@check_main_window_exists 
def WsClear_super_GUI() -> int:
	main_window.clear_workspace()
	return 0

@register_super_function('WsSample')
@check_main_window_exists
def WsSample_super_GUI() -> int:
	main_window.open_change_variables_sample_dialog()
	return 0

@register_super_function('WsExtrapolate')
@check_main_window_exists
def WsExtrapolate_super_GUI() -> int:
	main_window.open_extrapolate_variables_dialog()
	return 0

@register_super_function('WsHtoL')
@check_main_window_exists
def WsHtoL_super_GUI() -> int:
	main_window.open_high_to_low_dialog()
	return 0

@register_super_function('WsLtoH')
@check_main_window_exists
def WsLtoH_super_GUI() -> int:
	main_window.open_low_to_high_dialog()
	return 0

@register_super_function('WsSeasonAdj')
@check_main_window_exists
def WsSeasonAdj_super_GUI() -> int:
	lec = ""
	dialog = DynAdjustmentDialog(lec, main_window)
	dialog.open()
	return 0

@register_super_function('WsTrend')
@check_main_window_exists
def WsTrend_super_GUI() -> int:
	main_window.open_trend_correction_dialog()
	return 0

@register_super_function('ModelSimulate')
@check_main_window_exists 
def ModelSimulate_super_GUI() -> int:
	main_window.open_compute_simulation_dialog()
	return 0

@register_super_function('ModelCompile')
@check_main_window_exists
def ModelCompile_super_GUI() -> int:
	main_window.open_compute_model_dialog()
	return 0

@register_super_function('IdtExecute')
@check_main_window_exists
def IdtExecute_super_GUI() -> int:
	main_window.open_compute_identities_dialog()
	return 0

@register_super_function('EqsEstimate')
@check_main_window_exists
def EqsEstimate_super_GUI() -> int:
	tab_widget: IodeTabWidget = main_window.tab_widget
	tab_widget.open_edit_dialog(IodeType.EQUATIONS)
	return 0

@register_super_function('XodeRuleImport')
@check_main_window_exists
def XodeRuleImport_super_GUI() -> int:
	main_window.open_import_variables_dialog()
	return 0

@register_super_function('ViewTable')
@check_main_window_exists
def ViewTable_super_GUI(table_name: str, generalized_sample: str, nb_decimals: int) -> int:
	if table_name not in tables:
		QMessageBox.warning(None, "WARNING", "Could not compute the table '" + table_name + 
							"' with sample '" + generalized_sample + "'")
		return 1
	
	try:
		view = ComputedTableDialog(table_name, generalized_sample, nb_decimals, parent=main_window)
		view.open()
		return 0
	except Exception as e:
		QMessageBox.critical(None, "ERROR", "An error occurred while trying to view the table '" +
											table_name + "' with sample '" + generalized_sample + 
											"':\n" + str(e))
		return 1
